src/tasks/provider_manager.rs

```diff
@@ -50,41 +91,253 @@ pub async fn pull(providers: &[BuiltinProvider], all: bool) -> Result<(), Error>
+
+    let response = octocrab._get(download_url).await?;
+    let response = octocrab.follow_location_to_data(response).await?;
+    let zip_bytes = hyper::body::to_bytes(response.into_body()).await?;
+
+    let mut archive = zip::ZipArchive::new(Cursor::new(zip_bytes))?;
+    for i in 0..archive.len() {
+        let mut file = archive.by_index(i)?;
+        let provider = Path::new(file.name())
+            .file_stem()
+            .ok_or_else(|| Error::Other {
+                message: "Cannot determine provider name from zipped artifact".to_owned(),
+            })?
+            .to_owned();
+        let mut buffer = Vec::new();
+        file.read_to_end(&mut buffer)?;
+        fs::write(get_provider_destination(&provider), buffer)?;
     }
+
     Ok(())
 }
 
-async fn fetch_provider(provider: BuiltinProvider) -> Result<(), Error> {
-    let wasm_dir = providers_wasm_dir()?;
-    let provider_name = provider.name();
-    let location_url =
-        format!("https://github.com/fiberplane/proxy/raw/main/providers/{provider_name}.wasm");
+async fn fetch_latest_artifact(octocrab: &Octocrab, branch: &str) -> Result<Artifact, Error> {
+    let artifacts = get_artifacts(octocrab, "fiberplane-providers").await?;
+    artifacts
+        .items
+        .iter()
+        .find(|artifact| match &artifact.workflow_run {
+            Some(run) => run.head_branch == branch,
+            None => false,
+        })
+        .ok_or(Error::ProvidersNotFound)
+        .cloned()
+}
 
-    info!("Fetching {provider_name} provider at {location_url}");
+async fn get_artifacts(octocrab: &Octocrab, artifact_name: &str) -> Result<Page<Artifact>, Error> {
+    octocrab
+        .get(
+            "/repos/fiberplane/providers/actions/artifacts",
+            Some(&ArtifactsParams {
+                name: artifact_name.to_owned(),
+            }),
+        )
+        .await
+        .map_err(Error::from)
+}
 
-    let bytes = reqwest::get(location_url).await?.bytes().await?;
+#[derive(Parser)]
+pub struct BuildProvidersArgs {
+    /// Keep debugging information in the built provider(s).
+    #[clap(short, long)]
+    debug: bool,
 
-    let target_path = wasm_dir.join(format!("{provider_name}.wasm"));
-    if target_path.exists() {
-        return Err(Error::NoOverwrite {
-            provider: provider_name,
-            path: target_path,
+    /// A specific provider to build. If omitted, all providers are built.
+    #[clap(long, default_value = "all")]
+    provider: String,
+
+    /// Path where the checkout of the `providers` repository can be found.
+    #[clap(long, default_value = "../providers")]
+    providers_dir: PathBuf,
+}
+
+pub fn build_providers(args: BuildProvidersArgs) -> Result<(), Error> {
+    ensure_wasm_dir()?;
+
+    let providers_dir = args.providers_dir;
+    if !providers_dir.exists() {
+        return Err(Error::Other {
+            message: format!("Providers dir `{}` doesn't exist", providers_dir.display()),
```

Comment:
I don't remember if I left that message like this in the beginning, but I think that error messages that eventually will be read by users should avoid abbreviations when possible

```suggestion
            message: format!("Providers directory `{}` doesn't exist", providers_dir.display()),
```